Another indexing problem ... "index based filter not matching real file filter"

I've turned on `paranoid` and `index-threshold=0` in the caldav functional tests now ... this stops me from doing a release, so I will have to revert it unless we get this issue resolved relatively fast :-)

I get this traceback:

```
E           caldav.lib.error.ReportError: ReportError at '500 Internal Server Error
E
E           b'500 Internal Server Error\n\nServer got itself in trouble'', reason no reason

caldav/objects.py:194: ReportError
---------------------------------------------------------- Captured log call -----------------------------------------------------------
ERROR    aiohttp.server:web_protocol.py:403 Error handling request
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 433, in _handle_request
    resp = await request_handler(request)
  File "/usr/lib/python3.10/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/home/tobias/caldav/tests/test_caldav.py", line 2199, in xandikos_handler
    return await mainapp.aiohttp_handler(request, "/")
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/webdav.py", line 2152, in aiohttp_handler
    response = await self._handle_request(request, environ)
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/webdav.py", line 2108, in _handle_request
    return await do.handle(request, environ, self)
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/webdav.py", line 1764, in handle
    return await reporter.report(
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/webdav.py", line 353, in wrapper
    async for resp in req_fn(self, environ, *args, **kwargs):
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/caldav.py", line 566, in report
    async for (href, resource) in webdav.traverse_resource(
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/webdav.py", line 1205, in traverse_resource
    for (child_name, child_resource) in members_fn(resource):
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/web.py", line 637, in calendar_query
    for (name, file, etag) in self.store.iter_with_filter(filter=filter):
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/store/__init__.py", line 351, in _iter_with_filter_indexes
    raise AssertionError(
AssertionError: index based filter not matching real file filter
```

All tests pass if I turn off paranoid or if I turn the index threshold high.

I realize that this is a bit too little information for further debugging, but ... I need to sleep now, will get back with more details tomorrow.

```
ERROR    aiohttp.server:web_protocol.py:403 Error handling request
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 433, in _handle_request
    resp = await request_handler(request)
  File "/usr/lib/python3.10/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/home/tobias/caldav/tests/test_caldav.py", line 2199, in xandikos_handler
    return await mainapp.aiohttp_handler(request, "/")
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/webdav.py", line 2152, in aiohttp_handler
    response = await self._handle_request(request, environ)
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/webdav.py", line 2108, in _handle_request
    return await do.handle(request, environ, self)
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/webdav.py", line 1764, in handle
    return await reporter.report(
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/webdav.py", line 353, in wrapper
    async for resp in req_fn(self, environ, *args, **kwargs):
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/caldav.py", line 566, in report
    async for (href, resource) in webdav.traverse_resource(
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/webdav.py", line 1205, in traverse_resource
    for (child_name, child_resource) in members_fn(resource):
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/web.py", line 637, in calendar_query
    for (name, file, etag) in self.store.iter_with_filter(filter=filter):
  File "/usr/lib/python3.10/site-packages/xandikos-0.2.8-py3.10.egg/xandikos/store/__init__.py", line 352, in _iter_with_filter_indexes
    raise AssertionError(
AssertionError: index based filter CalendarFilter([ComponentFilter('VCALENDAR', children=[ComponentFilter('VTODO', children=[PropertyFilter('COMPLETED', children=[], is_not_defined=True, time_range=None), PropertyFilter('STATUS', children=[TextMatcher('STATUS', 'COMPLETED', collation=<function <lambda> at 0x7fcad093f9a0>, negate_condition=True)], is_not_defined=False, time_range=None), PropertyFilter('STATUS', children=[TextMatcher('STATUS', 'CANCELLED', collation=<function <lambda> at 0x7fcad093f9a0>, negate_condition=True)], is_not_defined=False, time_range=None)], is_not_defined=False, time_range=None)], is_not_defined=False, time_range=None)]) (values: {'C=VCALENDAR/C=VTODO/P=COMPLETED': [b'20221118T065947Z'], 'C=VCALENDAR/C=VTODO/P=STATUS': [b'NEEDS-ACTION', b'COMPLETED', b'NEEDS-ACTION']}) not matching real file filter
=================================================
```

--- a/xandikos/icalendar.py
+++ b/xandikos/icalendar.py
@@ -946,11 +946,11 @@ class CalendarFilter(Filter):
                 return False
         return True
 
-    def index_keys(self) -> list[str]:
-        subindexes = []
+    def index_keys(self) -> list[list[str]]:
+        result = []
         for child in self.children:
-            subindexes.extend(child.index_keys())
-        return subindexes
+            result.extend(child.index_keys())
+        return result
 
     def __repr__(self) -> str:
         return f"{self.__class__.__name__}({self.children!r})"

--- a/xandikos/store/__init__.py
+++ b/xandikos/store/__init__.py
@@ -158,7 +158,7 @@ class Filter:
         """
         raise NotImplementedError(self.check)
 
-    def index_keys(self) -> list[IndexKey]:
+    def index_keys(self) -> list[list[IndexKey]]:
         """Returns a list of indexes that could be used to apply this filter.
 
         Returns: AND-list of OR-options

--- a/xandikos/store/index.py
+++ b/xandikos/store/index.py
@@ -96,7 +96,7 @@ class AutoIndexManager:
         self.indexing_threshold = threshold
 
     def find_present_keys(
-        self, necessary_keys: Iterable[IndexKey]
+        self, necessary_keys: Iterable[Iterable[IndexKey]]
     ) -> Optional[Iterable[IndexKey]]:
         available_keys = self.index.available_keys()
         needed_keys = []

--- a/xandikos/tests/test_store.py
+++ b/xandikos/tests/test_store.py
@@ -23,6 +23,8 @@ import shutil
 import stat
 import tempfile
 import unittest
+from zoneinfo import ZoneInfo
+
 
 from dulwich.objects import Blob, Commit, Tree
 from dulwich.repo import Repo
@@ -36,7 +38,7 @@ from xandikos.store import (
     Store,
 )
 
-from ..icalendar import ICalendarFile
+from ..icalendar import ICalendarFile, CalendarFilter
 from ..store.git import BareGitStore, GitStore, TreeGitStore
 from ..store.vdir import VdirStore
 
@@ -465,3 +467,96 @@ class ExtractRegularUIDTests(unittest.TestCase):
     def test_extract_no_uid(self):
         fi = File([EXAMPLE_VCALENDAR_NO_UID], "text/bla")
         self.assertRaises(NotImplementedError, fi.get_uid)
+
+
+class ParanoidModeTests(unittest.TestCase):
+    """Test for the original issue #235: AssertionError in paranoid mode with index_threshold=0."""
+
+    def test_index_keys_return_type_bug(self):
+        """Test that demonstrates the index_keys() return type bug directly."""
+        # Create a filter with multiple components/properties to trigger the extend() bug
+        filter = CalendarFilter(ZoneInfo("UTC"))
+        component_filter = filter.filter_subcomponent("VCALENDAR")
+        todo_filter = component_filter.filter_subcomponent("VTODO")
+        todo_filter.filter_property("SUMMARY")
+        todo_filter.filter_property("CREATED")
+
+        index_keys = filter.index_keys()
+
+        # The fixed code should return a list of lists (AND-list of OR-options)
+        # Each inner list represents OR-options for a single AND requirement
+        self.assertIsInstance(index_keys, list)
+        if index_keys:
+            # Each element should be a list of strings (OR-options)
+            for key_group in index_keys:
+                self.assertIsInstance(
+                    key_group,
+                    list,
+                    f"Expected list of lists, but got list containing {type(key_group)} in {index_keys}",
+                )
+                for key in key_group:
+                    self.assertIsInstance(
+                        key,
+                        str,
+                        f"Expected strings in inner lists, but got {type(key)} in {key_group}",
+                    )
+
+        # Expected structure with the fix: [['C=VCALENDAR/C=VTODO/P=SUMMARY'], ['C=VCALENDAR/C=VTODO/P=CREATED']]
+        expected_keys = [
+            ["C=VCALENDAR/C=VTODO/P=SUMMARY"],
+            ["C=VCALENDAR/C=VTODO/P=CREATED"],
+        ]
+        self.assertEqual(index_keys, expected_keys)
+
+    def test_paranoid_mode_index_threshold_zero(self):
+        """Test that reproduces the original AssertionError when using paranoid mode with index_threshold=0.
+
+        This test verifies the fix for: "AssertionError: index based filter not matching real file filter"
+        which occurred when double_check_indexes=True and index_threshold=0.
+        """
+        # Create a temporary store with paranoid mode enabled and index_threshold=0
+        d = tempfile.mkdtemp()
+        self.addCleanup(shutil.rmtree, d)
+
+        store_path = os.path.join(d, "store")
+        os.mkdir(store_path)
+        repo = Repo.init_bare(store_path)
+        repo._autogc_disabled = True
+
+        store = BareGitStore(
+            repo,
+            double_check_indexes=True,  # Enable paranoid mode
+            index_threshold=0,  # Force immediate indexing
+        )
+        store.load_extra_file_handler(ICalendarFile)
+
+        # Ensure repository is properly closed after test
+        self.addCleanup(repo.close)
+
+        # Import a calendar file
+        (name, etag) = store.import_one(
+            "test.ics", "text/calendar", [EXAMPLE_VCALENDAR1]
+        )
+
+        # Create a CalDAV filter that would trigger the bug
+        # The old bug happened when index_keys() returned list[str] instead of list[list[str]]
+        # causing find_present_keys to iterate over characters instead of keys
+        filter = CalendarFilter(ZoneInfo("UTC"))
+        component_filter = filter.filter_subcomponent("VCALENDAR")
+        todo_filter = component_filter.filter_subcomponent("VTODO")
+        todo_filter.filter_property("SUMMARY")
+
+        # Force a scenario that uses the index by pre-indexing and then filtering
+        # First, let the filter create some index entries
+        results = list(store.iter_with_filter(filter))
+
+        # The old code would sometimes work on first run but fail on subsequent runs
+        # when the index state differed. Let's try again to trigger the comparison failure.
+        results2 = list(store.iter_with_filter(filter))
+
+        # Verify we get the expected results
+        self.assertEqual(len(results), 1)
+        self.assertEqual(len(results2), 1)
+        result_name, result_file, result_etag = results[0]
+        self.assertEqual(result_name, name)
+        self.assertEqual(result_etag, etag)

--- a/xandikos/tests/test_vcard.py
+++ b/xandikos/tests/test_vcard.py
@@ -158,7 +158,7 @@ class CardDAVFilterTests(unittest.TestCase):
         )  # Should not be in index keys
 
         keys = filter.index_keys()
-        self.assertEqual(sorted(keys), ["P=EMAIL", "P=FN"])
+        self.assertEqual(sorted(keys), [["P=EMAIL"], ["P=FN"]])
 
     def test_check_from_indexes(self):
         """Test checking from indexes."""

--- a/xandikos/vcard.py
+++ b/xandikos/vcard.py
@@ -328,13 +328,13 @@ class CardDAVFilter(Filter):
 
         return self.test(results) if results else True
 
-    def index_keys(self) -> list[str]:
+    def index_keys(self) -> list[list[str]]:
         """Return the index keys needed for this filter."""
-        keys = []
+        result = []
         for prop_filter in self.property_filters:
             if not prop_filter.is_not_defined:
-                keys.append(f"P={prop_filter.name}")
-        return keys
+                result.append([f"P={prop_filter.name}"])
+        return result
 
 
 def parse_filter(filter_el, cls):

